Add new `onRequest`/`onResponse` methods to `useUpstream`

hanford opened the pull request:
This PR adds two new lifecycle methods to useUpstream, `onRequest` and `onResponse`. These allow you to modify the `request` and `response` objects before they're returned, allowing the ability to set headers (programmatically) and also rewrite URLs before the request happens

```ts
reflare.push({
  path: '/*',

  onResponse: maybeAddCacheHeaders,
  onRequest: rewritePath,

  upstream: {
    domain: 'example.com',
    protocol: 'https',
  },
});

...

function rewritePath (req: Request, url: string): Request {
  return new Request(url.replace('/original/request/path', ''), req);
}

function maybeAddCacheHeaders(res: Response, url: string): Response {
  if (url.endsWith('.html') || url.endsWith('/')) {
    res.headers.set('cache-control', 'public, max-age=240, s-maxage=60');
  }

  return res;
}
```

Not sure if we should have these at the top level or inside of the `upstream` config, I think it makes sense either way and happy to adjust my PR if you'd like me to change it.

Refs #371 
Refs #352

xiaoyang-sde commented on types/index.ts:
```diff
@@ -14,6 +14,8 @@ export interface Route {
   headers?: HeadersOptions;
   methods?: string[],
   loadBalancing?: LoadBalancingOptions,
+  onResponse?: (k: Response, url: string) => Response;
+  onRequest?: (k: Request, url: string) => Request;
```
```suggestion
  onRequest?: (request: Request, url: string) => Request;
  onResponse?: (response: Response, url: string) => Response;
```

xiaoyang-sde commented on types/middleware.ts:
```diff
@@ -7,6 +7,8 @@ export interface Context {
   request: Request;
   response: Response;
   upstream: UpstreamOptions | null;
+  onResponse?: (k: Response, url: string) => Response;
+  onRequest?: (k: Request, url: string) => Request;
```
```suggestion
  onRequest?: (request: Request, url: string) => Request;
  onResponse?: (response: Response, url: string) => Response;
```

xiaoyang-sde reviewed (commented):
LGTM! Thanks for the contribution. Actually I prefer having these functions inside the `upstream` config, because they will be invoked inside the `upstream` middleware.

hanford commented:
@xiaoyang-sde I've moved the new methods into upstream options and added a little information about them to the README

Happy to make any other changes if you'd like!

xiaoyang-sde commented:
> 

Thanks! I will take another look later today.
